File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/InitialGpsEventFactory.java
package com.mapbox.services.android.navigation.v5.navigation;

import com.mapbox.core.utils.TextUtils;

class InitialGpsEventFactory {

  private static final String EMPTY_STRING = "";
  private String sessionId = EMPTY_STRING;
  private ElapsedTime time;
  private InitialGpsEventHandler handler;
  private boolean hasSent;
  private NavigationPerformanceMetadata metadata;

  InitialGpsEventFactory(NavigationPerformanceMetadata metadata) {
    this(new ElapsedTime(), new InitialGpsEventHandler(), metadata);
  }

  InitialGpsEventFactory(ElapsedTime time, InitialGpsEventHandler handler,
                         NavigationPerformanceMetadata metadata) {
    this.time = time;
    this.handler = handler;
    this.metadata = metadata;
  }

  void navigationStarted(String sessionId) {
    this.sessionId = sessionId;
    time.start();
  }

  void gpsReceived() {
    if (time.getStart() == null) {
      return;
    }
    time.end();
    send(time);
  }

  void reset() {
    time = new ElapsedTime();
    hasSent = false;
  }

  private void send(ElapsedTime time) {
    if (!hasSent && !TextUtils.isEmpty(sessionId)) {
      double elapsedTime = time.getElapsedTime();
      handler.send(elapsedTime, sessionId, metadata);
      hasSent = true;
    }
  }
}

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/InitialGpsEventFactory.kt
package com.mapbox.services.android.navigation.v5.navigation

import com.mapbox.core.utils.TextUtils

private const val EMPTY_STRING = ""

internal class InitialGpsEventFactory @JvmOverloads constructor(private var time: ElapsedTime = ElapsedTime(), private val handler: InitialGpsEventHandler = InitialGpsEventHandler()) {
    private var sessionId = EMPTY_STRING
    private var hasSent: Boolean = false

    fun navigationStarted(sessionId: String) {
        this.sessionId = sessionId
        time.start()
    }

    fun gpsReceived() {
        if (time.start == null) {
            return
        }
        time.end()
        send(time)
    }

    fun reset() {
        time = ElapsedTime()
        hasSent = false
    }

    private fun send(time: ElapsedTime) {
        if (!hasSent && !TextUtils.isEmpty(sessionId)) {
            val elapsedTime = time.elapsedTime
            handler.send(elapsedTime, sessionId)
            hasSent = true
        }
    }

}
